Add bLUSD yield strategies

edmulraney opened the pull request:
Adds the following yield strategies for bLUSD (from Liquity's LUSD chicken bonds protocol: chickenbonds.org):

**bLUSD rebond strategy**
'Rebonding bLUSD strategy continuously performs the following steps: create an LUSD bond, claim it at the optimum rebond time, sell the acquired bLUSD back to LUSD, and then bond again.'

**bLUSD/LUSD-3CRV LP strategy**
'Staking bLUSD/LUSD-3CRV LP tokens in the Curve gauge earns trade fees, Curve rewards and LUSD rewards from claimed bonds.'

llamatester commented:
The lusd-chickenbonds adapter exports pools: 
        
 
 Test Suites: 1 passed, 1 total
    Tests:       18 passed, 18 total
    Snapshots:   0 total
    Time:        0.322 s
    Ran all test suites.
    
    Nb of pools: 2
     
    
    Sample pools:
    ┌─────────┬──────────────────────────────────────────────┬─────────────────────┬───────────────────┬────────────┬───────────────────┬────────────────────┬──────────────────────────────────┬───────────────────────────────────────────────────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───────────────────┐
    │ (index) │                     pool                     │       project       │      symbol       │   chain    │      tvlUsd       │      apyBase       │         underlyingTokens         │                   rewardTokens                    │                                                                                          poolMeta                                                                                          │     apyReward     │
    ├─────────┼──────────────────────────────────────────────┼─────────────────────┼───────────────────┼────────────┼───────────────────┼────────────────────┼──────────────────────────────────┼───────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼───────────────────┤
    │    0    │ '0x57619FE9C539f890b19c61812226F9703ce37137' │ 'lusd-chickenbonds' │      'bLUSD'      │ 'ethereum' │ 48533773.61608891 │ 39.58860551894139  │       [ 'LUSD', 'bLUSD' ]        │                    [ 'LUSD' ]                     │ 'Rebonding bLUSD strategy continuously performs the following steps: create an LUSD bond, claim it at the optimum rebond time, sell the acquired bLUSD back to LUSD, and then bond again.' │                   │
    │    1    │ '0x74ED5d42203806c8CDCf2F04Ca5F60DC777b901c' │ 'lusd-chickenbonds' │ 'bLUSD/LUSD-3CRV' │ 'ethereum' │ 5075198.30921106  │ 0.4456241440979669 │ [ 'LUSD', 'bLUSD', 'LUSD-3CRV' ] │ [ 'bLUSD', 'LUSD-3CRV', 'LUSD', ... 1 more item ] │                           'Staking bLUSD/LUSD-3CRV LP tokens in the Curve gauge earns yield from trade fees, Curve rewards and LUSD rewards from claimed bonds.'                           │ 6.944300150883639 │
    └─────────┴──────────────────────────────────────────────┴─────────────────────┴───────────────────┴────────────┴───────────────────┴────────────────────┴──────────────────────────────────┴───────────────────────────────────────────────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────────────┘

llamatester commented:
The lusd-chickenbonds adapter exports pools: 
        
 
 Test Suites: 1 passed, 1 total
    Tests:       18 passed, 18 total
    Snapshots:   0 total
    Time:        0.319 s
    Ran all test suites.
    
    Nb of pools: 2
     
    
    Sample pools:
    ┌─────────┬──────────────────────────────────────────────┬─────────────────────┬───────────────────┬────────────┬───────────────────┬────────────────────┬──────────────────────────────────┬───────────────────────────────────────────────────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───────────────────┐
    │ (index) │                     pool                     │       project       │      symbol       │   chain    │      tvlUsd       │      apyBase       │         underlyingTokens         │                   rewardTokens                    │                                                                                          poolMeta                                                                                          │     apyReward     │
    ├─────────┼──────────────────────────────────────────────┼─────────────────────┼───────────────────┼────────────┼───────────────────┼────────────────────┼──────────────────────────────────┼───────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼───────────────────┤
    │    0    │ '0x57619FE9C539f890b19c61812226F9703ce37137' │ 'lusd-chickenbonds' │      'bLUSD'      │ 'ethereum' │ 51546188.64586131 │ 29.75992056796737  │       [ 'LUSD', 'bLUSD' ]        │                    [ 'LUSD' ]                     │ 'Rebonding bLUSD strategy continuously performs the following steps: create an LUSD bond, claim it at the optimum rebond time, sell the acquired bLUSD back to LUSD, and then bond again.' │                   │
    │    1    │ '0x74ED5d42203806c8CDCf2F04Ca5F60DC777b901c' │ 'lusd-chickenbonds' │ 'bLUSD/LUSD-3CRV' │ 'ethereum' │ 4616534.499037882 │ 0.3974546086117803 │ [ 'LUSD', 'bLUSD', 'LUSD-3CRV' ] │ [ 'bLUSD', 'LUSD-3CRV', 'LUSD', ... 1 more item ] │                           'Staking bLUSD/LUSD-3CRV LP tokens in the Curve gauge earns yield from trade fees, Curve rewards and LUSD rewards from claimed bonds.'                           │ 4.770550855162995 │
    └─────────┴──────────────────────────────────────────────┴─────────────────────┴───────────────────┴────────────┴───────────────────┴────────────────────┴──────────────────────────────────┴───────────────────────────────────────────────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────────────┘

slasher125 commented:
hey, ty for adding. sorry for the late reply.

`underlyingTokens` and `rewardTokens` should be addresses, not symbols.

edmulraney commented:
Thanks @slasher125 - token symbols are now replaced with addresses

slasher125 commented:
> Thanks @slasher125 - token symbols are now replaced with addresses

sorry i missed this one reeee. ty for making the change!

llamatester commented:
Error while running lusd-chickenbonds adapter: 
        
 
 Test Suites: 1 failed, 1 total
    Tests:       1 failed, 17 passed, 18 total
    Snapshots:   0 total
    Time:        0.36 s
    Ran all test suites.
    
    Nb of pools: 2
     
    
    Sample pools:
    ┌─────────┬──────────────────────────────────────────────┬─────────────────────┬───────────────────┬────────────┬───────────────────┬────────────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───────────────────┐
    │ (index) │                     pool                     │       project       │      symbol       │   chain    │      tvlUsd       │      apyBase       │                                                               underlyingTokens                                                               │                                                                         rewardTokens                                                                          │                                                                                          poolMeta                                                                                          │     apyReward     │
    ├─────────┼──────────────────────────────────────────────┼─────────────────────┼───────────────────┼────────────┼───────────────────┼────────────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼───────────────────┤
    │    0    │ '0x57619FE9C539f890b19c61812226F9703ce37137' │ 'lusd-chickenbonds' │      'bLUSD'      │ 'ethereum' │ 55021968.72846646 │ 28.39176397459915  │                        [ '0x5f98805A4E8be255a32880FDeC7F6728C6568bA0', '0xB9D7DdDca9a4AC480991865EfEf82E01273F79C3' ]                        │                                                       [ '0x5f98805A4E8be255a32880FDeC7F6728C6568bA0' ]                                                        │ 'Rebonding bLUSD strategy continuously performs the following steps: create an LUSD bond, claim it at the optimum rebond time, sell the acquired bLUSD back to LUSD, and then bond again.' │                   │
    │    1    │ '0x74ED5d42203806c8CDCf2F04Ca5F60DC777b901c' │ 'lusd-chickenbonds' │ 'bLUSD/LUSD-3CRV' │ 'ethereum' │ 4451353.483493235 │ 1.3609053125509307 │ [ '0x5f98805A4E8be255a32880FDeC7F6728C6568bA0', '0xB9D7DdDca9a4AC480991865EfEf82E01273F79C3', '0xEd279fDD11cA84bEef15AF5D39BB4d4bEE23F0cA' ] │ [ '0xB9D7DdDca9a4AC480991865EfEf82E01273F79C3', '0xEd279fDD11cA84bEef15AF5D39BB4d4bEE23F0cA', '0x5f98805A4E8be255a32880FDeC7F6728C6568bA0', ... 1 more item ] │                           'Staking bLUSD/LUSD-3CRV LP tokens in the Curve gauge earns yield from trade fees, Curve rewards and LUSD rewards from claimed bonds.'                           │ 8.504690757545227 │
    └─────────┴──────────────────────────────────────────────┴─────────────────────┴───────────────────┴────────────┴───────────────────┴────────────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────────────┘
    }

edmulraney commented:
> - there is an undefined in the token array which is breaking the test

Thanks, fixed.

>- lets remove the poolMeta fields (or trimm a lot: we usually use this field for fee tiers etc but not for proper descriptions as it would be way to long for the UI (its displayed right after the symbol on our frontend...)

Removed. Is there a way for DeFi Llama to show what the yield strategy is tho? bLUSD has a different APRs depending on which strategy its used in. Would be good for users to be able to understand which yield strategy is being referred to.

>- how come the apy is "off" compared to what i find on the different frontends. its displaying ~60%

They are two different APRs, bLUSD backing price APR vs bLUSD rebond strategy APR. The frontends don't currently show the rebonding APR but will be updated soon to include it.

The APR shown on frontends is the APR that the Treasury generates for the bLUSD backing price. It uses Yearn's LUSD Stability Pool APY to estimate the APY of the Stability Pool, however Yearn have now updated that strategy to include lending LUSD in AAVE, so we can no longer use it as an estimate of the bLUSD hold APR. We are looking into other ways to get the APY and will update our frontends. Once we've done that we'll add it to Defi Llama too but for now we've left it out.

llamatester commented:
The lusd-chickenbonds adapter exports pools: 
        
 
 Test Suites: 1 passed, 1 total
    Tests:       18 passed, 18 total
    Snapshots:   0 total
    Time:        0.316 s
    Ran all test suites.
    
    Nb of pools: 2
     
    
    Sample pools:
    ┌─────────┬──────────────────────────────────────────────┬─────────────────────┬───────────────────┬────────────┬────────────────────┬─────────────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───────────────────┐
    │ (index) │                     pool                     │       project       │      symbol       │   chain    │       tvlUsd       │       apyBase       │                                                               underlyingTokens                                                               │                                                                         rewardTokens                                                                          │     apyReward     │
    ├─────────┼──────────────────────────────────────────────┼─────────────────────┼───────────────────┼────────────┼────────────────────┼─────────────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼───────────────────┤
    │    0    │ '0x57619FE9C539f890b19c61812226F9703ce37137' │ 'lusd-chickenbonds' │      'bLUSD'      │ 'ethereum' │ 55536226.697695866 │ 24.817410843288258  │                        [ '0x5f98805A4E8be255a32880FDeC7F6728C6568bA0', '0xB9D7DdDca9a4AC480991865EfEf82E01273F79C3' ]                        │                                                       [ '0x5f98805A4E8be255a32880FDeC7F6728C6568bA0' ]                                                        │                   │
    │    1    │ '0x74ED5d42203806c8CDCf2F04Ca5F60DC777b901c' │ 'lusd-chickenbonds' │ 'bLUSD/LUSD-3CRV' │ 'ethereum' │  4487218.89296696  │ 0.05239452846235171 │ [ '0x5f98805A4E8be255a32880FDeC7F6728C6568bA0', '0xB9D7DdDca9a4AC480991865EfEf82E01273F79C3', '0xEd279fDD11cA84bEef15AF5D39BB4d4bEE23F0cA' ] │ [ '0xB9D7DdDca9a4AC480991865EfEf82E01273F79C3', '0xEd279fDD11cA84bEef15AF5D39BB4d4bEE23F0cA', '0x5f98805A4E8be255a32880FDeC7F6728C6568bA0', ... 1 more item ] │ 9.564429947912227 │
    └─────────┴──────────────────────────────────────────────┴─────────────────────┴───────────────────┴────────────┴────────────────────┴─────────────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────────────┘
